packetbeat/statsdns/statistics_dns.go

```diff
@@ -101,62 +101,80 @@ var (
 	IpsServer                    []string
 	UrlAnnouncementDeployFromBam string
 	MapViewIPs                   map[int]map[string][]string
+	QStatDNS                     *QueueStatDNS
+	IsActive                     bool
+	SubActive                    bool
 )
 
 func InitStatisticsDNS() {
 	GetConfigDNSStatistics()
+	// Create chan for management Statistic DNS counter
+	QStatDNS = NewQueueStatDNS()
+	IsActive := true
+	SubActive := false
+
 	go func() {
 		ticker := time.NewTicker(StatInterval * time.Second)
 		defer ticker.Stop()
-		StatSrv = &StatisticsService{Start: time.Now(), StatsMap: make(map[string]*StatisticsDNS, MaximumClients)}
-		ReqMap = &RequestMap{RequestMessage: make(map[string]map[string]string, MaximumClients)}
-		//===================Create Counter For PerView============================
-		//go func() {
-		//	CreateCounterMetricPerView(MapViewIPs)
-		//}()
-		//===================End Create Counter For PerView========================
-		for {
-			//===================Create Counter For PerView==========================
-			go func() {
-				CreateCounterMetricPerView(MapViewIPs)
-			}()
-			//===================End Create Counter For PerView======================
-			t := <-ticker.C
+		go func() {
+			QStatDNS.isActive = true
+			QStatDNS.SubStatDNS(&SubActive)
+		}()
+
+		for IsActive {
+			timeStart := time.Now()
+			logp.Info("Starting %s", timeStart)
+			StatSrv = &StatisticsService{Start: timeStart, StatsMap: make(map[string]*StatisticsDNS, MaximumClients)}
+			ReqMap = &RequestMap{RequestMessage: make(map[string]map[string]string, MaximumClients)}
+
+			CreateCounterMetricPerView(MapViewIPs)
+
+			// Active flag sub for counter
+			SubActive = true
+
+			timeEnd := <-ticker.C
 			mutex.Lock()
-			logp.Info("Starting %s", t)
-			StatSrv.End = t
+			SubActive = false
+			StatSrv.End = timeEnd
 			b, err := json.Marshal(StatSrv)
 			if err != nil {
 				logp.Error(err)
 				continue
 			}
+
 			logp.Info("DNS_Statistics: %s", b)
 			// open new thread to call the API
 			go func() {
 				outstats.PublishToSNMPAgent(string(b))
```

Comment:
Open an endless goroutine in a for loop like this might lead to high CPU and out of memory. You might want to use worker-pool pattern to limit the number of workers running at the same time.